python/marvin/contrib/vacs/base.py

```diff
@@ -0,0 +1,172 @@
+        * Retrieve the VAC data matching ``parent_object``. Usually one will
+          use attributes in ``parent_object`` such as ``.mangaid`` or
+          ``.plate`` to perform the match.
+        * Return the VAC data in whatever format is appropriate.
+
+        """
+
+        pass
+
+    @staticmethod
+    def get_vacs(parent_object):
+        """Returns a container with all the VACs subclassing from `VACMixIn`.
+
+        Because this method loops over ``VACMixIn.__subclasses__()``, all the
+        class that inherit from `VACMixIn` and that must be included in the
+        container need to have been imported before calling
+        `~VACMixIn.get_vacs`.
+
+        Parameters
+        ----------
+        parent_object : object
+            The object to which the VACs are being attached. It will be passed
+            to `~VACMixIn.get_data` when the subclass of `VACMixIn` is
+            called.
+
+        Returns
+        -------
+        vac_container : object
+            An instance of a class that contains just a list of properties, one
+            for to each on of the VACs that subclass from `VACMixIn`.
+
+        """
+
+        class VACContainer(object):
+
+            def __repr__(self):
+                return '<VACContainer ({0})>'.format(', '.join(map(repr, list(self))))
+
+            def __getitem__(self, value):
+                return getattr(self, value)
+
+            def __iter__(self):
+                for value in self.__dir__():
+                    if not value.startswith('_'):
+                        yield value
+
+        vac_container = VACContainer()
+
+        for subvac in VACMixIn.__subclasses__():
+            # We need to set sv=subvac in the lambda function to prevent
+            # a cell-var-from-loop issue.
+            setattr(VACContainer, subvac.name,
+                    property(lambda self, sv=subvac: sv().get_data(parent_object)))
+
+        return vac_container
+
+    def download_vac(self, path_params={}, verbose=True):
+        """Download the VAC using rsync and returns the local path."""
+
+        assert self.name in self.rsync_access.templates, 'VAC path has not been set in the tree.'
+
+        if verbose:
+            warnings.warn('downloading file for VAC {0!r}'.format(self.name), MarvinUserWarning)
+
+        self.rsync_access.remote()
+        self.rsync_access.add(self.name, **path_params)
+        self.rsync_access.set_stream()
+        self.rsync_access.commit()
+        paths = self.rsync_access.get_paths()
+
+        # adding a millisecond pause for download to finish and file existence to register
+        time.sleep(0.001)
+
+        return paths[0]  # doing this for single files, may need to change
+
+    def set_sandbox_path(self, relative_path):
+        """Downloads a file using its path relative to ``$MANGA_SANDBOX``."""
+
+        self.rsync_access.templates[self.name] = os.path.join('$MANGA_SANDBOX', relative_path)
+
+    def get_path(self, path_params={}):
```

Comment:
Additionally I think we don't need this as all officially released VACS live on the SAS in one official place, the SANDBOX which sdss_access already knows about it.  Marvin will only support officially approved and released VACs.

Reply:
Yeah, this is one of the things we should talk about. In general I agree, but I'm fairly concerned that having to create a PR in tree for a new path, understand the functionality there, etc may increase the steepness of contributing a VAC. I'm also not 100% sure we want to overcrowd the tree with VACs, many of which have filename format that are not 100% easy to implement with the current framework. What I was trying is to provide a simple way to use sdss_access to download a given path from the SAS without having to permanently add it to the tree. Anyway, let's discuss this on Monday.

Reply:
Since VACS only come around during DRs, and only in handfuls, I think we can get away with having one branch of tree called vacs and have people commit into that.  They are only adding one line in one line.  Given everything else they have to do for their VAC in general, I think that's ok.  

sdss_access isn't just for official sdss data products.  It's for every single file housed on the SAS; every file that has a datamodel html page.  If you look through the ini file, you'll find tons of intermediate and non-official files.  Plus, VACS have to create a datamodel html file as part of their release, and they become permanently housed by the SAS upon release, so I think it's fine to add this step in there for them.  If they have unconventional naming formats, I think that's even a stronger case to put it in sdss_access.

Reply:
That's a fair point. I want to make implementation of VACs in Marvin as easy as possible but make part of the VAC owner responsibilities before release to add a path in `sdss_paths.ini` seems like a good idea.